hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CellFlatMap.java

```diff
@@ -0,0 +1,468 @@
+
+    while (begin <= end) {
+      int mid = begin + ((end - begin) / 2);
+      Cell midCell = getCellFromIndex(mid);
+      int compareRes = comparator.compare(midCell, needle);
+
+      if (compareRes == 0) return mid;  // 0 means equals. We found the key
+      else if (compareRes < 0) {
+        // midCell is less than needle so we need to look at farther up
+        begin = mid + 1;
+      } else {
+        // midCell is greater than needle so we need to look down
+        end = mid - 1;
+      }
+    }
+
+    return (-1 * begin);
+  }
+
+  /* Get the index of the key taking into consideration whether
+  ** the key should be inclusive or exclusive */
+  private int getValidIndex(Cell key, boolean inclusive) {
+    int index = find(key);
+    if (inclusive && index >= 0)
+      index = (descending) ? index-1 : index+1;
+    return Math.abs(index);
+  }
+
+  @Override
+  public Comparator<? super Cell> comparator() {
+    return comparator;
+  }
+
+  @Override
+  public int size() {
+    return maxCellIdx-minCellIdx;
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return (maxCellIdx==minCellIdx);
+  }
+
+
+  // ---------------- Sub-Maps ----------------
+  @Override
+  public ConcurrentNavigableMap<Cell, Cell> subMap( Cell fromKey,
+                                                    boolean fromInclusive,
+                                                    Cell toKey,
+                                                    boolean toInclusive) {
+    int toIndex = getValidIndex(toKey, toInclusive);
+    int fromIndex = (getValidIndex(fromKey, !fromInclusive));
+
+    if (fromIndex > toIndex) throw new IllegalArgumentException("inconsistent range");
+    return createCellFlatMap(comparator, fromIndex, toIndex, descending);
+  }
+
+  @Override
+  public ConcurrentNavigableMap<Cell, Cell> headMap(Cell toKey, boolean inclusive) {
+    int index = getValidIndex(toKey, inclusive);
+    return createCellFlatMap(comparator, minCellIdx, index, descending);
+  }
+
+  @Override
+  public ConcurrentNavigableMap<Cell, Cell> tailMap(Cell fromKey, boolean inclusive) {
+    int index = (getValidIndex(fromKey, !inclusive));
+    return createCellFlatMap(comparator, index, maxCellIdx, descending);
+  }
+
+  @Override
+  public ConcurrentNavigableMap<Cell, Cell> descendingMap() {
+    return createCellFlatMap(comparator, minCellIdx, maxCellIdx, true);
+  }
+
+  @Override
+  public ConcurrentNavigableMap<Cell, Cell> subMap(Cell k1, Cell k2) {
+    return this.subMap(k1, true, k2, true);
+  }
+
+  @Override
+  public ConcurrentNavigableMap<Cell, Cell> headMap(Cell k) {
+    return this.headMap(k, true);
+  }
+
+  @Override
+  public ConcurrentNavigableMap<Cell, Cell> tailMap(Cell k) {
+    return this.tailMap(k, true);
+  }
+
+
+  // -------------------------------- Key's getters --------------------------------
+  @Override
+  public Cell firstKey() {
+    if (isEmpty()) return null;
+    if (descending) getCellFromIndex(maxCellIdx-1);
```

Comment:
Indeed, missing "return" statement. This code is generally not in use, but some times used for debugging and logging.